From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 30 Apr 2024 15:28:45 +0300
Subject: [PATCH] Item Modifiers API

Introduces the concept of "item modifiers". They are primary used to modify item packets before sending them to the player.
Item modifiers are a set of rules to apply to the item stacks before displaying them to the player.

The order of modifiers is handled by a string list of modifier ids in config (option "item-modifiers.modification-order").
Also adds "item-modifiers.add-unknown" to automatically add new modifier ids to the config
and "item-modifiers.log-missing-ids" to log if the id is missing from the config.

Plugins need to implement ItemModifier and call #register on them to enable them.

diff --git a/src/main/java/me/sosedik/kiterino/inventory/InventorySlotHelper.java b/src/main/java/me/sosedik/kiterino/inventory/InventorySlotHelper.java
new file mode 100644
index 0000000000000000000000000000000000000000..3d596bbfaa0015a5d2e34207f4065d9ad38f2344
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/inventory/InventorySlotHelper.java
@@ -0,0 +1,47 @@
+package me.sosedik.kiterino.inventory;
+
+/**
+ * Small helpers for working with inventory slots
+ */
+public final class InventorySlotHelper {
+
+    private InventorySlotHelper() {
+        throw new IllegalStateException("Utility class");
+    }
+
+    /**
+     * Item held in cursor
+     */
+    public static final int CURSOR = -1;
+
+    /**
+     * The head inventory slot
+     */
+    public static final int HEAD_SLOT = 5;
+
+    /**
+     * The chest/chestplate inventory slot
+     */
+    public static final int CHEST_SLOT = 6;
+
+    /**
+     * The legs/leggings inventory slot
+     */
+    public static final int LEGS_SLOT = 7;
+
+    /**
+     * The feet/boots inventory slot
+     */
+    public static final int FEET_SLOT = 8;
+
+    /**
+     * The first hotbar slot
+     */
+    public static final int FIRST_HOTBAR_SLOT = 36;
+
+    /**
+     * Item held in offhand
+     */
+    public static final int OFF_HAND = 45;
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/ItemContextBox.java b/src/main/java/me/sosedik/kiterino/modifier/item/ItemContextBox.java
new file mode 100644
index 0000000000000000000000000000000000000000..049ab18c07a36163f0dd13ac5acc628215bf4baf
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/ItemContextBox.java
@@ -0,0 +1,165 @@
+package me.sosedik.kiterino.modifier.item;
+
+import io.papermc.paper.datacomponent.DataComponentTypes;
+import io.papermc.paper.datacomponent.item.ItemLore;
+import me.sosedik.kiterino.modifier.item.context.ItemModifierContext;
+import me.sosedik.kiterino.modifier.item.context.ItemModifierContextType;
+import net.kyori.adventure.text.ComponentLike;
+import org.bukkit.Material;
+import org.bukkit.entity.Player;
+import org.bukkit.inventory.ItemStack;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+import java.util.List;
+import java.util.Locale;
+
+/**
+ * Contains data required for modifying items
+ */
+@SuppressWarnings("unused")
+@NullMarked
+public class ItemContextBox {
+
+    private final @Nullable Player viewer;
+    private final Locale locale;
+    private final ItemModifierContextType contextType;
+    private final ItemModifierContext context;
+    private final Material type;
+    private ItemStack item;
+
+    /**
+     * Creates a new context box with viewer's locale
+     *
+     * @param viewer player viewing the item
+     * @param contextType context type
+     * @param context modification context
+     * @param item item
+     */
+    public ItemContextBox(Player viewer, ItemModifierContextType contextType, ItemModifierContext context, ItemStack item) {
+        this(viewer, viewer.locale(), contextType, context, item);
+    }
+
+    /**
+     * Creates a new context box
+     *
+     * @param viewer player viewing the item
+     * @param locale locale for modifications
+     * @param contextType context type
+     * @param context modification context
+     * @param item item
+     */
+    public ItemContextBox(@Nullable Player viewer, Locale locale, ItemModifierContextType contextType, ItemModifierContext context, ItemStack item) {
+        this.viewer = viewer;
+        this.locale = locale;
+        this.contextType = contextType;
+        this.type = item.getType();
+        this.item = item;
+        this.context = context;
+    }
+
+    /**
+     * Gets the player receiving this item
+     *
+     * @return player viewer
+     */
+    public @Nullable Player getViewer() {
+        return this.viewer;
+    }
+
+    /**
+     * Locale that should be used for modification
+     *
+     * @return locale
+     */
+    public Locale getLocale() {
+        return this.locale;
+    }
+
+    /**
+     * Gets the context type
+     *
+     * @return context type
+     */
+    public ItemModifierContextType getContextType() {
+        return this.contextType;
+    }
+
+    /**
+     * Gets the modification context
+     *
+     * @return modification context
+     */
+    public ItemModifierContext getContext() {
+        return this.context;
+    }
+
+    /**
+     * Gets the original item's type
+     *
+     * @return item
+     */
+    public Material getInitialType() {
+        return this.type;
+    }
+
+    /**
+     * Sets the item's new type
+     */
+    public void setType(Material type) {
+        this.item = this.item.withType(type);
+    }
+
+    /**
+     * Gets the modifying item
+     *
+     * @return item
+     */
+    public ItemStack getItem() {
+        return this.item;
+    }
+
+    /**
+     * Completely replaces the item
+     *
+     * @param item item
+     */
+    public void setItem(ItemStack item) {
+        this.item = item;
+    }
+
+    /**
+     * Helper for adding a line to the item's lore
+     *
+     * @param line component
+     */
+    public void addLore(ComponentLike line) {
+        ItemLore lore;
+        if (this.item.hasData(DataComponentTypes.LORE)) {
+            ItemLore currentLore = this.item.getData(DataComponentTypes.LORE);
+            assert currentLore != null;
+            lore = ItemLore.lore().lines(currentLore.lines()).addLine(line).build();
+        } else {
+            lore = ItemLore.lore().addLine(line).build();
+        }
+        this.item.setData(DataComponentTypes.LORE, lore);
+    }
+
+    /**
+     * Helper for adding lines to the item's lore
+     *
+     * @param lines components
+     */
+    public void addLore(List<? extends ComponentLike> lines) {
+        ItemLore lore;
+        if (this.item.hasData(DataComponentTypes.LORE)) {
+            ItemLore currentLore = this.item.getData(DataComponentTypes.LORE);
+            assert currentLore != null;
+            lore = ItemLore.lore().lines(currentLore.lines()).addLines(lines).build();
+        } else {
+            lore = ItemLore.lore().addLines(lines).build();
+        }
+        this.item.setData(DataComponentTypes.LORE, lore);
+    }
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifier.java b/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifier.java
new file mode 100644
index 0000000000000000000000000000000000000000..fb6c78fb256d83c472ac504a524a08d953b8d348
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifier.java
@@ -0,0 +1,112 @@
+package me.sosedik.kiterino.modifier.item;
+
+import me.sosedik.kiterino.modifier.item.context.ItemModifierContext;
+import me.sosedik.kiterino.modifier.item.context.ItemModifierContextType;
+import org.bukkit.NamespacedKey;
+import org.bukkit.entity.Player;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Locale;
+
+/**
+ * Makes modifications to an item
+ */
+public abstract class ItemModifier {
+
+    private final NamespacedKey modifierId;
+
+    protected ItemModifier(@NotNull NamespacedKey modifierId) {
+        this.modifierId = modifierId;
+    }
+
+    /**
+     * Unique ID of this modifier
+     *
+     * @return modifier id
+     */
+    public @NotNull NamespacedKey getModifierId() {
+        return modifierId;
+    }
+
+    /**
+     * Registers this modifier
+     */
+    public void register() {
+        ItemModifiersHandler.itemModifiersHandler().registerModifier(this);
+    }
+
+    /**
+     * Unregisters this modifier
+     */
+    public void unregister() {
+        ItemModifiersHandler.itemModifiersHandler().unregisterModifier(this);
+    }
+
+    /**
+     * Whether this modifier should automatically
+     * skip air items
+     *
+     * @return whether to skip air items
+     */
+    public boolean skipAir() {
+        return true;
+    }
+
+    /**
+     * Whether this modifier should automatically
+     * skip the provided context
+     *
+     * @param contextType context type
+     * @return whether to skip context
+     */
+    public boolean skipContext(@NotNull ItemModifierContextType contextType) {
+        return false;
+    }
+
+    /**
+     * Modifies an item given the context
+     *
+     * @param contextBox item data
+     * @return modification result
+     */
+    public abstract @NotNull ModificationResult modify(@NotNull ItemContextBox contextBox);
+
+    /**
+     * Modify item with viewer's locale, empty context, lore and 0 slot.
+     * Will return {@code null} if no modifications were made.
+     *
+     * @param viewer player viewing the item
+     * @param item item
+     * @return modified item or {@code null}
+     */
+    public static @Nullable ItemStack modifyItem(@NotNull Player viewer, @NotNull ItemStack item) {
+        return modifyItem(viewer, viewer.locale(), item);
+    }
+
+    /**
+     * Modify item with empty context, lore and 0 slot
+     *
+     * @param viewer player viewing the item
+     * @param locale locale
+     * @param item item
+     * @return modified item or {@code null}
+     */
+    public static @Nullable ItemStack modifyItem(@Nullable Player viewer, @NotNull Locale locale, @NotNull ItemStack item) {
+        var contextBox = new ItemContextBox(viewer, locale, ItemModifierContextType.EMPTY_LORE, ItemModifierContext.EMPTY, item.clone());
+        return modifyItem(contextBox);
+    }
+
+    /**
+     * Modify item given the context.
+     * Will return {@code null} if no modifications were made.
+     *
+     * @param contextBox item context box
+     * @return modified item or {@code null}
+     */
+    public static @Nullable ItemStack modifyItem(@NotNull ItemContextBox contextBox) {
+        return ItemModifiersHandler.itemModifiersHandler().modifyItem(contextBox);
+    }
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandler.java b/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandler.java
new file mode 100644
index 0000000000000000000000000000000000000000..3f25b5678b9b44c69543bfe8a773bf4ac379c315
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandler.java
@@ -0,0 +1,45 @@
+package me.sosedik.kiterino.modifier.item;
+
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+@ApiStatus.Internal
+abstract class ItemModifiersHandler {
+
+    static ItemModifiersHandler itemModifiersHandler;
+
+    /**
+     * Registers item modifier
+     *
+     * @param itemModifier item modifier
+     */
+    public abstract void registerModifier(@NotNull ItemModifier itemModifier);
+
+    /**
+     * Unregisters item modifier
+     *
+     * @param itemModifier item modifier
+     */
+    public abstract void unregisterModifier(@NotNull ItemModifier itemModifier);
+
+    /**
+     * Applies modifiers to the item.
+     * Will return {@code null} if no changes were made.
+     *
+     * @param contextBox context box
+     * @return modified item or {@code null} if no changes
+     */
+    public abstract @Nullable ItemStack modifyItem(@NotNull ItemContextBox contextBox);
+
+    /**
+     * Gets the item modifiers applier
+     *
+     * @return item modifiers handler
+     */
+    public static @NotNull ItemModifiersHandler itemModifiersHandler() {
+        return itemModifiersHandler;
+    }
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/ModificationResult.java b/src/main/java/me/sosedik/kiterino/modifier/item/ModificationResult.java
new file mode 100644
index 0000000000000000000000000000000000000000..4b6555df17fc0b91a0f362e0f006f8c231a5f558
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/ModificationResult.java
@@ -0,0 +1,26 @@
+package me.sosedik.kiterino.modifier.item;
+
+/**
+ * Represents the result of modifying an item
+ */
+public enum ModificationResult {
+
+    /**
+     * Modified item
+     */
+    OK,
+    /**
+     * Skipped item
+     */
+    PASS,
+    /**
+     * Completely replaced item.
+     * This will skip any further modifiers.
+     * <p>
+     * Generally meant to be used by the highest priority modifiers
+     * that then redirect to {@link ItemModifier#modifyItem(ItemContextBox)}
+     * to go through the remaining modifiers.
+     */
+    RETURN
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/context/ItemModifierContext.java b/src/main/java/me/sosedik/kiterino/modifier/item/context/ItemModifierContext.java
new file mode 100644
index 0000000000000000000000000000000000000000..ab1ae078428eca044a7b908f081f5f7956eaf807
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/context/ItemModifierContext.java
@@ -0,0 +1,13 @@
+package me.sosedik.kiterino.modifier.item.context;
+
+/**
+ * Context within which the item is modified
+ */
+public interface ItemModifierContext {
+
+    /**
+     * Empty item modifier context
+     */
+    ItemModifierContext EMPTY = new ItemModifierContext() {};
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/context/ItemModifierContextType.java b/src/main/java/me/sosedik/kiterino/modifier/item/context/ItemModifierContextType.java
new file mode 100644
index 0000000000000000000000000000000000000000..9f1c536f5b9c366f91534866dd86425a3dd10305
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/context/ItemModifierContextType.java
@@ -0,0 +1,169 @@
+package me.sosedik.kiterino.modifier.item.context;
+
+import me.sosedik.kiterino.modifier.item.context.packet.AdvancementPacketContext;
+import me.sosedik.kiterino.modifier.item.context.packet.EntityDataPacketContext;
+import me.sosedik.kiterino.modifier.item.context.packet.EntityEquipmentPacketContext;
+import me.sosedik.kiterino.modifier.item.context.packet.MerchantOfferPacketContext;
+import me.sosedik.kiterino.modifier.item.context.packet.RecipeBookPacketContext;
+import me.sosedik.kiterino.modifier.item.context.packet.UnknownEntityDataContext;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Context type
+ */
+public class ItemModifierContextType {
+
+    /**
+     * Setting singular item in an inventory
+     */
+    public static final ItemModifierContextType SET_SLOT = context(SlottedItemModifierContext.class).withName().withLore().build();
+
+    /**
+     * Setting multiple items in an inventory
+     */
+    public static final ItemModifierContextType WINDOW_ITEMS = context(SlottedItemModifierContext.class).withName().withLore().build();
+
+    /**
+     * Equipment slots of other entities
+     */
+    public static final ItemModifierContextType ENTITY_EQUIPMENT = context(EntityEquipmentPacketContext.class).build();
+
+    /**
+     * Item inside the recipe book
+     */
+    public static final ItemModifierContextType RECIPE_BOOK = context(RecipeBookPacketContext.class).withName().withLore().build();
+
+    /**
+     * Ghost item as a response for the recipe book
+     */
+    public static final ItemModifierContextType RECIPE_GHOST = context(RecipeBookPacketContext.class).withName().withLore().build();
+
+    /**
+     * Advancement's item (also used in toasts due to lack of distinction)
+     */
+    public static final ItemModifierContextType ADVANCEMENT = context(AdvancementPacketContext.class).build();
+
+    /**
+     * Merchant offer (trade screen)
+     */
+    public static final ItemModifierContextType MERCHANT_OFFER = context(MerchantOfferPacketContext.class).withName().withLore().build();
+
+    /**
+     * Entity data packet
+     */
+    public static final ItemModifierContextType ENTITY_DATA = context(EntityDataPacketContext.class).build();
+
+    /**
+     * Unknown entity data packet
+     */
+    public static final ItemModifierContextType UNKNOWN_ENTITY_DATA = context(UnknownEntityDataContext.class).build();
+
+    /**
+     * Empty modification context with visible name and lore
+     */
+    public static final ItemModifierContextType EMPTY_LORE = context(ItemModifierContext.EMPTY.getClass()).withName().withLore().build();
+
+    /**
+     * Empty modification context with visible name and without visible lore
+     */
+    public static final ItemModifierContextType EMPTY_NO_LORE = context(ItemModifierContext.EMPTY.getClass()).withName().build();
+
+    private final Class<? extends ItemModifierContext> contextClass;
+    private final boolean visibleName;
+    private final boolean visibleLore;
+
+    /**
+     * Constructs a context type
+     *
+     * @param contextClass context class
+     * @param visibleName whether the name is visible in this context
+     * @param visibleLore whether the lore is visible in this context
+     */
+    public ItemModifierContextType(@NotNull Class<? extends ItemModifierContext> contextClass, boolean visibleName, boolean visibleLore) {
+        this.contextClass = contextClass;
+        this.visibleName = visibleName;
+        this.visibleLore = visibleLore;
+    }
+
+    /**
+     * Gets the context class
+     *
+     * @return context class
+     */
+    public @NotNull Class<? extends ItemModifierContext> getContextClass() {
+        return this.contextClass;
+    }
+
+    /**
+     * Whether the name is visible in this context
+     *
+     * @return whether the name is visible
+     */
+    public boolean hasVisibleName() {
+        return this.visibleName;
+    }
+
+    /**
+     * Whether the lore is visible in this context
+     *
+     * @return whether the lore is visible
+     */
+    public boolean hasVisibleLore() {
+        return this.visibleLore;
+    }
+
+    /**
+     * Constructs a builder for item modifier context type
+     *
+     * @param contextClass context class
+     * @return builder
+     */
+    public static @NotNull Builder context(@NotNull Class<? extends ItemModifierContext> contextClass) {
+        return new Builder(contextClass);
+    }
+
+    /**
+     * Item modifier context type builder
+     */
+    public static class Builder {
+
+        private final Class<? extends ItemModifierContext> contextClass;
+        private boolean visibleName;
+        private boolean visibleLore;
+
+        Builder(@NotNull Class<? extends ItemModifierContext> contextClass) {
+            this.contextClass = contextClass;
+        }
+
+        /**
+         * Marks that this context had a visible name
+         *
+         * @return this builder
+         */
+        public @NotNull Builder withName() {
+            this.visibleName = true;
+            return this;
+        }
+
+        /**
+         * Marks that this context had a visible lore
+         *
+         * @return this builder
+         */
+        public @NotNull Builder withLore() {
+            this.visibleLore = true;
+            return this;
+        }
+
+        /**
+         * Builds a context type
+         *
+         * @return context type
+         */
+        public @NotNull ItemModifierContextType build() {
+            return new ItemModifierContextType(this.contextClass, this.visibleName, this.visibleLore);
+        }
+
+    }
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/context/PacketItemModifierContext.java b/src/main/java/me/sosedik/kiterino/modifier/item/context/PacketItemModifierContext.java
new file mode 100644
index 0000000000000000000000000000000000000000..67f0cf829a98756d1196ad5227540828a4345d69
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/context/PacketItemModifierContext.java
@@ -0,0 +1,17 @@
+package me.sosedik.kiterino.modifier.item.context;
+
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Item modifier context that has an NMS packet
+ */
+public interface PacketItemModifierContext extends ItemModifierContext {
+
+    /**
+     * The item packet that was in the base of this context
+     *
+     * @return the item packet, if present
+     */
+    @NotNull Object packet();
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/context/SlottedItemModifierContext.java b/src/main/java/me/sosedik/kiterino/modifier/item/context/SlottedItemModifierContext.java
new file mode 100644
index 0000000000000000000000000000000000000000..cd932d04ae657c77a6947ec9bbe5e3d5f1c136f9
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/context/SlottedItemModifierContext.java
@@ -0,0 +1,18 @@
+package me.sosedik.kiterino.modifier.item.context;
+
+import me.sosedik.kiterino.inventory.InventorySlotHelper;
+
+/**
+ * Item modifier context that has a container slot
+ */
+public interface SlottedItemModifierContext extends ItemModifierContext {
+
+    /**
+     * Gets the raw container slot the item's in
+     *
+     * @return slot
+     * @see InventorySlotHelper
+     */
+    int slot();
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/AdvancementPacketContext.java b/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/AdvancementPacketContext.java
new file mode 100644
index 0000000000000000000000000000000000000000..7432bdba6c37eac184051ef5003270aedfc284ca
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/AdvancementPacketContext.java
@@ -0,0 +1,51 @@
+package me.sosedik.kiterino.modifier.item.context.packet;
+
+import me.sosedik.kiterino.modifier.item.context.PacketItemModifierContext;
+import org.bukkit.NamespacedKey;
+import org.bukkit.advancement.Advancement;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.Supplier;
+
+public class AdvancementPacketContext implements PacketItemModifierContext {
+
+    private final Object packet;
+    private final NamespacedKey advancementKey;
+    private final Supplier<Advancement> advancementSupplier;
+    private @Nullable Advancement advancement;
+
+    public AdvancementPacketContext(@NotNull Object packet, @NotNull NamespacedKey advancementKey, @NotNull Supplier<Advancement> advancementSupplier) {
+        this.packet = packet;
+        this.advancementKey = advancementKey;
+        this.advancementSupplier = advancementSupplier;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public @NotNull Object packet() {
+        return this.packet;
+    }
+
+    /**
+     * Gets the advancement key
+     *
+     * @return the advancement key
+     */
+    public @NotNull NamespacedKey getAdvancementKey() {
+        return this.advancementKey;
+    }
+
+    /**
+     * Fetches the bukkit advancement wrapper
+     *
+     * @return advancement
+     */
+    public @NotNull Advancement getAdvancement() {
+        if (this.advancement == null) this.advancement = advancementSupplier.get();
+        return this.advancement;
+    }
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/EntityDataPacketContext.java b/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/EntityDataPacketContext.java
new file mode 100644
index 0000000000000000000000000000000000000000..804e6d866a36cf76f8e9121c4c2ecf7dab8b2841
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/EntityDataPacketContext.java
@@ -0,0 +1,43 @@
+package me.sosedik.kiterino.modifier.item.context.packet;
+
+import org.bukkit.World;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.EntityType;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.Function;
+
+/**
+ * Wraps around entity data packet
+ */
+public class EntityDataPacketContext extends EntityPacketContext {
+
+    static Function<Object, @Nullable Entity> entityFetcher;
+
+    private final EntityType entityType;
+
+    /**
+     * Constructs wrapper around entity data packet
+     *
+     * @param packet nms packet
+     * @param world world
+     * @param entityId entity id
+     * @param entity entity instance
+     */
+    public EntityDataPacketContext(@NotNull Object packet, @NotNull World world, int entityId, @NotNull EntityType entityType, @NotNull Object entity) {
+        super(packet, world, entityId, () -> entityFetcher.apply(entity));
+        this.entityType = entityType;
+    }
+
+    /**
+     * Gets the entity type of this entity.
+     * Does not require fetching the entity instance.
+     *
+     * @return entity type
+     */
+    public @NotNull EntityType getEntityType() {
+        return this.entityType;
+    }
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/EntityEquipmentPacketContext.java b/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/EntityEquipmentPacketContext.java
new file mode 100644
index 0000000000000000000000000000000000000000..e85d7bb003b14a3eeca800c1a43bbcbf69b69285
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/EntityEquipmentPacketContext.java
@@ -0,0 +1,39 @@
+package me.sosedik.kiterino.modifier.item.context.packet;
+
+import org.bukkit.World;
+import org.bukkit.entity.Entity;
+import org.bukkit.inventory.EquipmentSlot;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Wrapper around entity equipment packet
+ */
+public class EntityEquipmentPacketContext extends EntityPacketContext {
+
+    private final EquipmentSlot slot;
+
+    /**
+     * Constructs entity equipment packet wrapper
+     *
+     * @param packet nms packet
+     * @param world entity's world
+     * @param entityId internal entity id
+     * @param entity entity
+     * @param slot equipment slot
+     */
+    public EntityEquipmentPacketContext(@NotNull Object packet, @NotNull World world, int entityId, @Nullable Entity entity, @NotNull EquipmentSlot slot) {
+        super(packet, world, entityId, entity);
+        this.slot = slot;
+    }
+
+    /**
+     * Gets the equipment slot
+     *
+     * @return the equipment slot
+     */
+    public @NotNull EquipmentSlot getSlot() {
+        return slot;
+    }
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/EntityPacketContext.java b/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/EntityPacketContext.java
new file mode 100644
index 0000000000000000000000000000000000000000..78b066594ca3c5bdb7a3721448c9bee8be463a06
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/EntityPacketContext.java
@@ -0,0 +1,102 @@
+package me.sosedik.kiterino.modifier.item.context.packet;
+
+import me.sosedik.kiterino.modifier.item.context.PacketItemModifierContext;
+import org.bukkit.World;
+import org.bukkit.entity.Entity;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.Supplier;
+
+/**
+ * Wrapper around packet containing an entity
+ */
+public abstract class EntityPacketContext implements PacketItemModifierContext {
+
+    private final Object packet;
+    private final World world;
+    private final int entityId;
+    protected @Nullable Entity entity;
+    private @Nullable Supplier<@Nullable Entity> entityFetcher;
+    protected boolean fetchedEntity;
+
+    /**
+     * Constructs wrapper around packet that contains an entity
+     *
+     * @param packet nms packet
+     * @param world world instance
+     * @param entityId internal entity id
+     * @param entity entity
+     */
+    protected EntityPacketContext(@NotNull Object packet, @NotNull World world, int entityId, @Nullable Entity entity) {
+        this.packet = packet;
+        this.world = world;
+        this.entityId = entityId;
+        this.entity = entity;
+        this.fetchedEntity = true;
+    }
+
+    /**
+     * Constructs wrapper around packet that contains an entity
+     *
+     * @param packet nms packet
+     * @param world world instance
+     * @param entityId internal entity id
+     * @param entityFetcher entity supplier
+     */
+    protected EntityPacketContext(@NotNull Object packet, @NotNull World world, int entityId, @Nullable Supplier<@Nullable Entity> entityFetcher) {
+        this.packet = packet;
+        this.world = world;
+        this.entityId = entityId;
+        this.entityFetcher = entityFetcher;
+        this.fetchedEntity = entityFetcher == null;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public @NotNull Object packet() {
+        return this.packet;
+    }
+
+    /**
+     * Gets the internal entity id
+     *
+     * @return the entity id
+     */
+    public int getEntityId() {
+        return this.entityId;
+    }
+
+    /**
+     * Get the world the entity's in
+     *
+     * @return world
+     */
+    public @NotNull World getWorld() {
+        return this.world;
+    }
+
+    /**
+     * Fetches the Bukkit entity by its internal id
+     *
+     * @return bukkit entity
+     */
+    public @Nullable Entity getEntity() {
+        if (this.fetchedEntity) return this.entity;
+        this.fetchedEntity = true;
+        if (entityFetcher != null) this.entity = entityFetcher.get();
+        return this.entity;
+    }
+
+    /**
+     * Checks whether the entity was fetched
+     *
+     * @return whether the entity was fetched
+     */
+    public boolean isFetched() {
+        return this.fetchedEntity;
+    }
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/MerchantOfferPacketContext.java b/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/MerchantOfferPacketContext.java
new file mode 100644
index 0000000000000000000000000000000000000000..6e59e0f8cabd301599bbf6f78747851d336991ea
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/MerchantOfferPacketContext.java
@@ -0,0 +1,38 @@
+package me.sosedik.kiterino.modifier.item.context.packet;
+
+import me.sosedik.kiterino.modifier.item.context.PacketItemModifierContext;
+import org.bukkit.inventory.MerchantRecipe;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Wraps around merchant offer packet
+ *
+ * @param packet nms packet
+ * @param merchantRecipe nms merchant offer
+ * @param slot item slot
+ */
+public record MerchantOfferPacketContext(@NotNull Object packet,
+                                         @NotNull MerchantRecipe merchantRecipe,
+                                         @NotNull Slot slot) implements PacketItemModifierContext {
+
+    /**
+     * Represents trade screen's item slot position
+     */
+    public enum Slot {
+
+        /**
+         * First item cost
+         */
+        FIRST,
+        /**
+         * Second item cost
+         */
+        SECOND,
+        /**
+         * Trade result
+         */
+        RESULT
+
+    }
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/RecipeBookPacketContext.java b/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/RecipeBookPacketContext.java
new file mode 100644
index 0000000000000000000000000000000000000000..be6268789b92aca7d10300f9f1829d20b4e251b9
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/RecipeBookPacketContext.java
@@ -0,0 +1,51 @@
+package me.sosedik.kiterino.modifier.item.context.packet;
+
+import me.sosedik.kiterino.modifier.item.context.PacketItemModifierContext;
+import org.bukkit.NamespacedKey;
+import org.jspecify.annotations.NullMarked;
+
+/**
+ * Recipe book context
+ *
+ * @param packet nms packet
+ * @param displayType item's display position
+ */
+@NullMarked
+public record RecipeBookPacketContext(Object packet,
+                                      DisplayType displayType) implements PacketItemModifierContext {
+
+	/**
+	 * Represents item's display position
+	 */
+	public enum DisplayType {
+		/**
+		 * Resulting item
+		 */
+		RESULT,
+		/**
+		 * Ingredient item
+		 */
+		INGREDIENT,
+		/**
+		 * Crafting station
+		 */
+		CRAFTING_STATION,
+		/**
+		 * Fuel item
+		 */
+		FUEL,
+		/**
+		 * Smithing template
+		 */
+		SMITHING_TEMPLATE,
+		/**
+		 * Smithing base
+		 */
+		SMITHING_BASE,
+		/**
+		 * Smithing addition
+		 */
+		SMITHING_ADDITION,
+	}
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/SlottedItemPacketContext.java b/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/SlottedItemPacketContext.java
new file mode 100644
index 0000000000000000000000000000000000000000..2499cf046a74c8604f1020c3115f61f8d099c21a
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/SlottedItemPacketContext.java
@@ -0,0 +1,12 @@
+package me.sosedik.kiterino.modifier.item.context.packet;
+
+import me.sosedik.kiterino.modifier.item.context.PacketItemModifierContext;
+import me.sosedik.kiterino.modifier.item.context.SlottedItemModifierContext;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Wrapper around set slot and window items packets
+ */
+public record SlottedItemPacketContext(@NotNull Object packet, int slot) implements PacketItemModifierContext, SlottedItemModifierContext {
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/UnknownEntityDataContext.java b/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/UnknownEntityDataContext.java
new file mode 100644
index 0000000000000000000000000000000000000000..e38ae04d3d13f3c86e7570589be02e8452ef620e
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/context/packet/UnknownEntityDataContext.java
@@ -0,0 +1,25 @@
+package me.sosedik.kiterino.modifier.item.context.packet;
+
+import org.bukkit.World;
+import org.bukkit.entity.Entity;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.function.Supplier;
+
+/**
+ * Wrapper for entity data packets that have entity id, but can't fetch the bukkit entity
+ */
+public class UnknownEntityDataContext extends EntityPacketContext {
+
+    /**
+     * Constructs wrapper around packet that contains an unknown entity
+     *
+     * @param packet   nms packet
+     * @param world    world instance
+     * @param entityId internal entity id
+     */
+    public UnknownEntityDataContext(@NotNull Object packet, @NotNull World world, int entityId) {
+        super(packet, world, entityId, (Supplier<Entity>) null);
+    }
+
+}
